### ISSUE TITLE
Dictionaries for new DQM Dataformat

### ISSUE THREAD

**Issue by schneiml** [cms-sw/cmssw#27277]
Dear framework experts,

the new DataFormat for DQM proposed in #26603 has evolved a bit over the last weeks and we got to actually using it as a product. The latest version can be found here:
https://github.com/schneiml/cmssw/blob/726d7c14e5c71977923c674674f64f84c041042f/DataFormats/Histograms/interface/MonitorElementCollection.h

Now edm tells me
```
No data dictionary found for the following classes:

  std::unique_ptr<MonitorElementData,std::default_delete<MonitorElementData> >
```

but trying to add an entry to `classes_def.xml` leads to the error
```
>> Building LCG reflex dict from header file src/DataFormats/Histograms/src/classes.h
Error: It is not necessary to explicitly select class unique_ptr<MonitorElementData,default_delete<MonitorElementData> >. I/O is supported for it transparently.
```

How to proceed? 

More specifically, which constraints apply to persistent and non-persistent products? For persistence, it seems copying needs to be supported, which makes the use of `unique_ptr` rather hard (and obviously the locks would need to be made transient somehow). There seem to be less constraints for non-persistent products, but still something is wrong here...

I can't see anything in my code that would be mentioned in the relevant TWiki page: https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCreatingNewProducts#Restrictions_on_the_package_defi 

For context, the branch https://github.com/cms-sw/cmssw/compare/master...schneiml:dqm-new-dqmstore-on-CMSSW_11_0_0_pre1 should apply cleanly on pre1 and contains some test code in the `DQMServices/Demo` package. However, not all of CMSSW compiles at the moment: only `DataFormats`, `DQMServices` and `DQM`.

@andrius-k @fioriNTU FYI

**Comment by Dr15Jones** [comment]
@pcanal suggestions?

**Comment by Dr15Jones** [comment]
@schneiml when you say 'Now edm tells me' what exactly tells you that? scram? cmsRun?

**Comment by pcanal** [comment]
There is no need for a dictionary for unique_ptr<T>.

Where does the error message comes from?  Which version of ROOT are you using?

We remove this (or similar) message from TClass::GetMissingDictionaries in v6.14/06 and v6.16/00.

**Comment by schneiml** [comment]
@Dr15Jones “edm tells me” == I do cmsrun and get [1].
@pcanal ROOT as provided by CMSSW 11 [2].

[1]
```
[mschneid][16][cmsdev32][.../src/DQMServices/Demo]$ (dqm-new-dqmstore-on-CMSSW_11_0_0_pre1 ~ ../Core/interface/DQMEDHarvester.h ) 
cmsRun test/rundemo_cfg.py 
++ starting: constructing source: EmptySource
++ finished: constructing source: EmptySource
++++ starting: constructing module with label 'e' id = 1
++++ finished: constructing module with label 'e' id = 1
++++ starting: constructing module with label 'harv3' id = 2
--- got a branch!
++++ finished: constructing module with label 'harv3' id = 2
----- Begin Fatal Exception 19-Jun-2019 15:46:24 CEST-----------------------
An exception of category 'DictionaryNotFound' occurred while
   [0] Constructing the EventProcessor
   [1] Calling ProductRegistry::initializeLookupTables
Exception Message:
No data dictionary found for the following classes:

  std::unique_ptr<MonitorElementData,std::default_delete<MonitorElementData> >

Most likely each dictionary was never generated, but it may
be that it was generated in the wrong package. Please add
(or move) the specification '<class name="whatever"/>' to
the appropriate classes_def.xml file along with any other
information needed there. For example, if this class has any
transient members, you need to specify them in classes_def.xml.
Also include the class header in classes.h

A type listed above might or might not be the same as a
type stored in the Event (or Lumi or Run). Instead it might
be the type of a data member, base class, wrapped type, or
other object needed by a stored type. Below is some additional
information which lists the stored types associated with the
types whose dictionaries were not found:

  MonitorElementCollection (produced in current process)

Missing dictionaries are associated with these branch names:

  MonitorElementCollection_harv3_DQMGenerationHarvestingLumi_TEST.
  MonitorElementCollection_harv3_DQMGenerationHarvestingRun_TEST.

If you do not need these branches and they are not produced
in the current process, an alternate solution to adding
dictionaries is to drop these branches on input using the
inputCommands parameter of the PoolSource.
----- End Fatal Exception -------------------------------------------------
[mschneid][16][cmsdev32][.../src/DQMServices/Demo]71$ (dqm-new-dqmstore-on-CMSSW_11_0_0_pre1 ~ ../Core/interface/DQMEDHarvester.h ) 
```

[2]

```
which root
/cvmfs/cms.cern.ch/slc7_amd64_gcc700/cms/cmssw/CMSSW_11_0_0_pre1/external/slc7_amd64_gcc700/bin/root
```

```
   ------------------------------------------------------------
  | Welcome to ROOT 6.14/09                http://root.cern.ch |
  |                               (c) 1995-2018, The ROOT Team |
  | Built for linuxx8664gcc                                    |
  | From tag , 22 November 2018                                |
  | Try '.help', '.demo', '.license', '.credits', '.quit'/'.q' |
   ------------------------------------------------------------
```

**Comment by Dr15Jones** [comment]
On a quick glance, `ProductRegistry::initializeLookupTables` calls  `TClass::GetMissingDictionaries` to get the list of missing dictionaries.

**Comment by pcanal** [comment]
I can not trivially reproduce the problem ...

**Comment by schneiml** [comment]
@pcanal "trivial" instructions for reproducing, copy&paste into a lxplus7 shell:
```
mkdir /dev/shm/$USER
cd /dev/shm/$USER
cmsrel CMSSW_11_0_0_pre1
cd CMSSW_11_0_0_pre1/src/
cmsenv
git cms-merge-topic schneiml:dqm-new-dqmstore-on-CMSSW_11_0_0_pre1
cd DataFormats/
scram b -j10
cd ../DQMServices/Demo/
scram b -j10
cmsRun test/rundemo_cfg.py 

```

Should run through in < 2minutes. Needs CMS stuff (cmsrel, cmsev etc.), obviously.

**Comment by pcanal** [comment]
The instructions worked to give me a reproducer.  Thanks.

**Comment by schneiml** [comment]
Some more experiments on my side show that the problem is not related to the `unique_ptr`; even without `unique_ptr` the problem hits on a plain `MonitorElementData*`. Changing the entries in `MonitorElementCollection` to values avoids the issue.

This is good enough for now, though it does make the object ownership rather messy and I'd really prefer pointers for readability as well as performance reasons.

**Comment by pcanal** [comment]
> On a quick glance, `ProductRegistry::initializeLookupTables` calls `TClass::GetMissingDictionaries` to get the list of missing dictionaries.

humm it seems to be using:
```
382                   if (!public_base_classes(missingDictionaries, containedTypeID, baseTypes)) {
```
which uses
```
322         if (!checkDictionary(missingDictionaries, typeID)) {
```
which is 
```
67        bool checkDictionary(std::vector<std::string>& missingDictionaries, TypeID const& typeID) {
68          TClass::GetClass(typeID.typeInfo());
69          if (!hasDictionary(typeID.typeInfo())) {
70            // a second attempt to load
71            TypeWithDict::byName(typeID.className());
72          }
(gdb) l
73          if (!hasDictionary(typeID.typeInfo())) {
74            missingDictionaries.emplace_back(typeID.className());
75            return false;
76          }
77          return true;
78        }
```
which does
```
814         return (TClassTable::GetDict(ti) != nullptr);
```
which indeed will always return a nullptr for std::unique_ptr<..>.

**Comment by pcanal** [comment]
So it looks like the code in either FWCore/Utilities/src/DictionaryTools.cc or DataFormats/Provenance/src/ProductRegistry.cc:382 needs to tweaked to look at the pointee whenever it sees a raw pointer or a unique_ptr.

**Comment by schneiml** [comment]
@Dr15Jones @pcanal can we get this fixed on CMSSW side?

**Comment by Dr15Jones** [comment]
@makortel could you look into this?

**Comment by makortel** [comment]
@Dr15Jones Will do.

**Comment by makortel** [comment]
I wonder if `TypeWithDict::byName()` would be a good place for stripping off the `unique_ptr`. It seems to be doing the job already for e.g. raw pointers
https://github.com/cms-sw/cmssw/blob/9ef01ce62c21a09cf614bc1b0d8263ce21251386/FWCore/Reflection/src/TypeWithDict.cc#L112-#L124

Do we have tests for `TypeWithDict`?

**Comment by schneiml** [comment]
FYI: I found a different, but related bug, reported over there: https://github.com/cms-sw/cmssw/issues/27291#issuecomment-505879046
I think actually I need that one fixed.

**Comment by makortel** [comment]
I have now fixed
* what @pcanal suggested in https://github.com/cms-sw/cmssw/issues/27277#issuecomment-504100606
* the underscore from `std::default_delete` @schneiml reported in https://github.com/cms-sw/cmssw/issues/27291#issuecomment-505879046

I'm testing producing `std::vector<std::unique_ptr<int>>` and have added dictionary to be generated for `edm::Wrapper<std::vector<std::unique_ptr<int>>>` and am now getting missing dictionary exception being thrown at
https://github.com/cms-sw/cmssw/blob/7488a8f9600316ad2f43b32f236c797af19573f4/FWCore/Framework/src/ProductRegistryHelper.cc#L115#-L118
after the call to
https://github.com/cms-sw/cmssw/blob/7488a8f9600316ad2f43b32f236c797af19573f4/FWCore/Reflection/src/DictionaryTools.cc#L139-#L165
Essentially this piece
```cpp
TClass* tClass = typeWithDict.getClass();
...
tClass->GetMissingDictionaries(hashTable, recursive);
```
adds `vector<unique_ptr<int,default_delete<int> > >` to the `hashTable` when the function `checkClassDictionaries()` is called with `edm::Wrapper<std::vector<std::unique_ptr<int,std::default_delete<int> > > >`.

@pcanal What should I do here? Use the pointed-to type here as well, i.e. obtain the `TClass` for `edm::Wrapper<std::vector<int> >` perhaps?

**Comment by makortel** [comment]
@schneiml 
> Edit: turns out @makortel hit that one before: [e13def5#diff-1cc804fec844be08287e9668b6889f10R16](https://github.com/cms-sw/cmssw/commit/e13def53a492d1d9dbca549d6e0ad94d482cb002#diff-1cc804fec844be08287e9668b6889f10R16)

Actually `std::shared_ptr` should work already (as does `std::unique_ptr` in some cases at least) since we do have a "friendly-name" translation for `std::shared_ptr` and `std::unique_ptr`. What you discovered in https://github.com/cms-sw/cmssw/issues/27291#issuecomment-505879046 is that the default template argument to the deleter of `std::unique_ptr`, i.e. `std::default_delete`, needs a special treatment as well.

**Comment by pcanal** [comment]
FYI, std::shared_ptr is not yet support by the I/O.

**Comment by pcanal** [comment]
> have added dictionary to be generated for edm::Wrapper<std::vector<std::unique_ptr<int>>>

In the generated source file, what is the spelling used?

**Comment by makortel** [comment]
@pcanal 
So in the `classes_def.xml` I added
```xml
<class name="edm::Wrapper<std::vector<std::unique_ptr<int>>>"/> 
```
and the `DataFormatsWrappedStdDictionaries_xr.cc` has e.g.
```cpp
namespace ROOT {
   static TClass *edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR_Dictionary();
   static void edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR_TClassManip(TClass*);
   static void *new_edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR(void *p = 0);
   static void *newArray_edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR(Long_t size, void *p);
   static void delete_edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR(void *p);
   static void deleteArray_edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR(void *p);
   static void destruct_edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR(void *p);

   // Function generating the singleton type initializer
   static TGenericClassInfo *GenerateInitInstanceLocal(const ::edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > >*)
   {
      ::edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > > *ptr = 0;
      static ::TVirtualIsAProxy* isa_proxy = new ::TIsAProxy(typeid(::edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > >));
      static ::ROOT::TGenericClassInfo 
         instance("edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > >", ::edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > >::Class_Version(), "DataFormats/Common/interface/Wrapper.h", 25,
                  typeid(::edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > >), ::ROOT::Internal::DefineBehavior(ptr, ptr),
                  &edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR_Dictionary, isa_proxy, 4,
                  sizeof(::edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > >) );
      instance.SetNew(&new_edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR);
      instance.SetNewArray(&newArray_edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR);
      instance.SetDelete(&delete_edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR);
      instance.SetDeleteArray(&deleteArray_edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR);
      instance.SetDestructor(&destruct_edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR);

      ::ROOT::AddClassAlternate("edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > >","edm::Wrapper<std::vector<std::unique_ptr<int> > >");
      return &instance;
   }
   TGenericClassInfo *GenerateInitInstance(const ::edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > >*)
   {
      return GenerateInitInstanceLocal((::edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > >*)0);
   }
   // Static variable to force the class initialization
   static ::ROOT::TGenericClassInfo *_R__UNIQUE_DICT_(Init) = GenerateInitInstanceLocal((const ::edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > >*)0x0); R__UseDummy(_R__UNIQUE_DICT_(Init));

   // Dictionary for non-ClassDef classes
   static TClass *edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR_Dictionary() {
      TClass* theClass =::ROOT::GenerateInitInstanceLocal((const ::edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > >*)0x0)->GetClass();
      edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR_TClassManip(theClass);
   return theClass;
   }

   static void edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR_TClassManip(TClass* ){
   }

} // end of namespace ROOT
```

**Comment by makortel** [comment]
I won't be able to work on this for several weeks, so here is the branch with fixes until 
https://github.com/cms-sw/cmssw/issues/27277#issuecomment-506760287
and further debugging
https://github.com/cms-sw/cmssw/compare/master...makortel:uniquePtrDictionary

**Comment by makortel** [comment]
(trying to resume)

I tested what I suggested in https://github.com/cms-sw/cmssw/issues/27277#issuecomment-506760287 i.e.
> Use the pointed-to type here as well, i.e. obtain the TClass for edm::Wrapper<std::vector<int> > perhaps?

in 
https://github.com/cms-sw/cmssw/blob/7488a8f9600316ad2f43b32f236c797af19573f4/FWCore/Reflection/src/DictionaryTools.cc#L139-L165

After that I get an exception from
https://github.com/cms-sw/cmssw/blob/be9bbe2726172744c09a7dac81a001bfe3f83300/FWCore/Reflection/src/TypeWithDict.cc#L368-L370
that is called from
https://github.com/cms-sw/cmssw/blob/be9bbe2726172744c09a7dac81a001bfe3f83300/DataFormats/Provenance/src/ProductRegistry.cc#L342
The exception says `No Dictionary for class: 'vector<unique_ptr<int,default_delete<int> > >'`.

@pcanal Am I going in the right direction? So far it feels that each patch reveals a next place to be patched. Should `TypeWithDict` itself be made aware of the special treatments for `unique_ptr`?

**Comment by pcanal** [comment]
```
The exception says No Dictionary for class: 'vector<unique_ptr<int,default_delete<int> > >'.
```
That's odd .. https://github.com/cms-sw/cmssw/issues/27277#issuecomment-506841849 says that there is TClass being generate(able) for that name:
```
instance("edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > >", ::edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > >::Class_Version(), "DataFormats/Common/interface/Wrapper.h", 25,
                  typeid(::edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > >), ::ROOT::Internal::DefineBehavior(ptr, ptr),
                  &edmcLcLWrapperlEvectorlEunique_ptrlEintcOdefault_deletelEintgRsPgRsPgRsPgR_Dictionary, isa_proxy, 4,
                  sizeof(::edm::Wrapper<vector<unique_ptr<int,default_delete<int> > > >) );

```
humm no ....
Did you try to generate the dictionary explicitly for ```vector<unique_ptr<int,default_delete<int> > >```?

**Comment by makortel** [comment]
@pcanal If I add
```xml
 <class name="std::vector<std::unique_ptr<int> >" />
```
(or with the explicit `default_delete<int>`) to `DataFormats/StdDictionaries/src/classes_def_vector.xml`, `scram b` fails with
```
>> Building LCG reflex dict from header file src/DataFormats/StdDictionaries/src/classes_vector.h
...
Warning: Unused class rule: std::vector<std::unique_ptr<int> >
gmake: *** [config/SCRAM/GMake/Makefile.rules:1732: tmp/slc7_amd64_gcc700/src/DataFormats/StdDictionaries/src/DataFormatsStdDictionaries/a/DataFormatsStdDictionaries_xr.cc] Error 1
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

**Comment by pcanal** [comment]
This 'usually' means that the genreflex invocation (and/or the classes.h) is missing headers that are needed for the template instantiation ....

**Comment by makortel** [comment]
Thanks @pcanal, the missing include was indeed the issue. That actually fixed a lot of things, and it turns out that what I had done until https://github.com/cms-sw/cmssw/issues/27277#issuecomment-506760287 is sufficient to support `vector<unique_ptr<T>>` (but not `unique_ptr<T>` alone).

(I took a quick look further why `unique_ptr<T>`, and the symptom is the same as in https://github.com/cms-sw/cmssw/issues/27277#issuecomment-522720425)

I'm thinking to make a PR for the `vector<unique_ptr<T>>` case now, leaving the `unique_ptr<T>` later (if it is really needed). Unless @pcanal or @Dr15Jones think that the remaining issue with `unique_ptr<T>` would point to something deeper that would be better fixed in some other way than what's needed for `vector<unique_ptr<T>>`.

**Comment by pcanal** [comment]
We implemented support for unique_ptr<T> by making it (mostly) transparent the the Meta and I/O layer in ROOT (and looking and treated like a T*) Thus (theoretically) there is never a need to request a dictionary for it.  However when it is used as a template parameter, you need to spelt it 'normally' (and the need for the dict for vector<unique_ptr<T>>)

**Comment by makortel** [comment]
I discussed with @pcanal yesterday and essentially he said that regarding ROOT we should treat `unique_ptr` as we treat raw pointers. Here is a sort-of a braindump of what I've experimented with

1. Change `edm::checkDictionary()` https://github.com/cms-sw/cmssw/blob/60c56004422a5145aa497e064781b077638e2872/FWCore/Reflection/src/DictionaryTools.cc#L69-L76 such that if the "second attempt to load" branch is taken, replace `std::unique_ptr<T, anything>` with `T`, call `TypeWithDict::byName` with the `T`, and use the `std::type_info` from the returned `TypeWithDict` for the final `hasDictionary()` call
   * This change is sufficient to get `std::vector<std::unique_ptr<int>>` produced by an EDProducer
2. Change `edm::TypeWithDict::byName()` such that it does the same for `unique_ptr<T>` as for `T*`
https://github.com/cms-sw/cmssw/blob/60c56004422a5145aa497e064781b077638e2872/FWCore/Reflection/src/TypeWithDict.cc#L112-L124
    * In order to get producing `std::vector<std::unique_ptr<int>>` working, I had call `TypeWithDict::toType()` to get a handle to the pointed-to type in two places
       * In `edm::checkDictionary()` change https://github.com/cms-sw/cmssw/blob/60c56004422a5145aa497e064781b077638e2872/FWCore/Reflection/src/DictionaryTools.cc#L71-L73 to essentially `hasDictionary(TypeWithDict::byName(typeID.className()).toType().typeInfo())`
       * In `ProductRegistry::initializeLookupTables()` change https://github.com/cms-sw/cmssw/blob/60c56004422a5145aa497e064781b077638e2872/DataFormats/Provenance/src/ProductRegistry.cc#L342 to `TypeID typeID(desc.unwrappedType().toType().typeInfo());`


I did the attempt 2 partly to experiment what would it mean to try to produce `unique_ptr<int>` from an EDProducer. The attempt 2 above is not sufficient though, because (at least) when setting the put token indices in `ProducerBase::resolvePutIndicies()` this condition https://github.com/cms-sw/cmssw/blob/60c56004422a5145aa497e064781b077638e2872/FWCore/Framework/src/ProducerBase.cc#L91-L92 the `*std::get<0>(it->second))` is `int` while `tl.typeID_` is `std::unique_ptr<int>`.

**Comment by makortel** [comment]
> I did the attempt 2 partly to experiment what would it mean to try to produce `unique_ptr<int>` from an EDProducer. The attempt 2 above is not sufficient though, because (at least) when setting the put token indices in `ProducerBase::resolvePutIndicies()` this condition
> 
> https://github.com/cms-sw/cmssw/blob/60c56004422a5145aa497e064781b077638e2872/FWCore/Framework/src/ProducerBase.cc#L91-L92
> 
> the `*std::get<0>(it->second))` is `int` while `tl.typeID_` is `std::unique_ptr<int>`.

I traced back the origin of `int` to my
> In `ProductRegistry::initializeLookupTables()` change https://github.com/cms-sw/cmssw/blob/60c56004422a5145aa497e064781b077638e2872/DataFormats/Provenance/src/ProductRegistry.cc#L342 to `TypeID typeID(desc.unwrappedType().toType().typeInfo());`

Here `desc`  is the `BranchDescription`, which internally holds the branch C++ type as `edm::TypeWithDict`. The `desc.unwrappedType()` returns the `edm::TypeWithDict` for `std::unique_ptr<int>`. However, internally it had transformed the type to `int *` (because we're in attempt 2). I added the `toType()` call here because otherwise I get an exception
```
Function TypeWithDict::typeInfo: Type
int*
does not have valid type_info information in ROOT
because it is a pointer.
```
from https://github.com/cms-sw/cmssw/blob/38b00fb65db80c9248afb64b78c13aadd3e86aaf/FWCore/Reflection/src/TypeWithDict.cc#L357-L378

Based on all this I'd say
* `edm::TypeWithDict` should distinguish between `T *` and `std::unique_ptr<T>`
* We can not support `T *` or `std::unique_ptr<T>` (or enum or C-array) products as of today
* In order to support `std::vector<std::unique_ptr<T>>`, maybe the attempt 1 of https://github.com/cms-sw/cmssw/issues/27277#issuecomment-523577726 is a feasible "minimal fix"

(unless we re-think how we propagate the type information through the system and check for existence of dictionaries)

**Comment by makortel** [comment]
After an intensive pair debugging session with @Dr15Jones we concluded that the proper fix is actually something completely different, and is implemented in #27861.

**Comment by schneiml** [comment]
Hi Matti,

I will try, though that involves undoing a bunch of workarounds all over the place, so it may take a bit. I'll report later.

Thanks anyways!
